sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs

```diff
@@ -0,0 +1,306 @@
+                    }
+                    else
+                    {
+                        OnCollectionAccessed(instruction, collectionSymbol.HasConstraint(CollectionConstraint.Empty, programState));
+                    }
+                }
+
+                return newProgramState;
+            }
+
+            private ProgramState ProcessInvocation(ProgramState programState, InvocationExpressionSyntax instruction)
+            {
+                var newProgramState = programState;
+
+                var memberAccess = instruction.Expression as MemberAccessExpressionSyntax;
+                if (memberAccess != null)
+                {
+                    var collectionSymbol = semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol;
+                    var collectionType = GetCollectionType(collectionSymbol);
+
+                    if (collectionType.IsAny(CollectionTypes))
+                    {
+                        var methodSymbol = semanticModel.GetSymbolInfo(instruction).Symbol as IMethodSymbol;
+                        if (methodSymbol == null ||
+                            methodSymbol.IsExtensionMethod ||
+                            IgnoredMethods.Contains(methodSymbol.Name))
+                        {
+                            return newProgramState;
+                        }
+
+                        if (AddMethods.Contains(methodSymbol.Name))
+                        {
+                            newProgramState = collectionSymbol.SetConstraint(CollectionConstraint.NotEmpty, newProgramState);
+                        }
+                        else
+                        {
+                            OnCollectionAccessed(instruction, collectionSymbol.HasConstraint(CollectionConstraint.Empty, programState));
+                        }
+                    }
+                }
+
+                newProgramState = RemoveCollectionConstraintsFromArguments(instruction, newProgramState);
+
+                return newProgramState;
+            }
+
+            private static ProgramState RemoveCollectionConstraintsFromArguments(InvocationExpressionSyntax instruction, ProgramState newProgramState)
+            {
+                var values = instruction.ArgumentList.Arguments.Select((node, index) =>
+                {
+                    SymbolicValue value;
+                    newProgramState.ExpressionStack.Pop(out value);
+                    return value;
+                });
+                foreach (var value in values)
+                {
+                    newProgramState = value.RemoveConstraint(CollectionConstraint.Empty, newProgramState);
+                }
+
+                return newProgramState;
+            }
+
+            public override ProgramState ObjectCreated(ProgramState programState, SymbolicValue symbolicValue, SyntaxNode instruction)
+            {
+                var newProgramState = programState;
+                switch (instruction.Kind())

```

Comment:
It might sounds dumb for you as you have worked on the rule but I would like to see some comments with the kind of case you handle with the `case` and `if` because it is sometimes hard to be sure what you are handling.